File: Aug/Day19/number-of-zero-filled-subarrays.cc
/* 
    Approach: Incremental counting approach
    Time: O(N)  ::  Space: O(1)
*/ 
class Solution {
public:
    long long zeroFilledSubarray(vector<int>& nums) {
        long long totalZeroFilledSubarrays = 0, currentZeroFilledSubarrays = 0;
        for (int num : nums) {
            if (num == 0) {
                currentZeroFilledSubarrays++;
                totalZeroFilledSubarrays += currentZeroFilledSubarrays;
            } else {
                currentZeroFilledSubarrays = 0;
            }
        }
        return totalZeroFilledSubarrays;
    }
};



/* 
    Approach: Arithmetic series formula
    Time: O(N)  ::  Space: O(1)
*/ 
class Solution {
public:
    long long zeroFilledSubarray(vector<int>& nums) {
        long long totalZeroFilledSubarrays = 0, currentZeroFilledSubarrays = 0;
        for (int num : nums) {
            if (num != 0) {
                long long cnt = (currentZeroFilledSubarrays * (currentZeroFilledSubarrays + 1)) / 2;
                totalZeroFilledSubarrays += cnt;
                currentZeroFilledSubarrays = 0;
            } else {
                currentZeroFilledSubarrays++;
            }
        }
        if (currentZeroFilledSubarrays != 0) {
            long long cnt = (currentZeroFilledSubarrays * (currentZeroFilledSubarrays + 1)) / 2;
            totalZeroFilledSubarrays += cnt;
        }
        return totalZeroFilledSubarrays;
    }
};

File: Aug/Day19/number-of-zero-filled-subarrays.py
#  Time: O(N)  ::  Space: O(1)
class Solution:
    def zeroFilledSubarray(self, nums: List[int]) -> int:
        totalZeroFilledSubarrays = currentZeroFilledSubarrays = 0

        for num in nums:
            if num == 0:
                currentZeroFilledSubarrays+=1
                totalZeroFilledSubarrays+=currentZeroFilledSubarrays
            else:
                currentZeroFilledSubarrays = 0
        return totalZeroFilledSubarrays
        



        
#  Time: O(N)  ::  Space: O(1)
class Solution:
    def zeroFilledSubarray(self, nums: List[int]) -> int:
        totalZeroFilledSubarrays = currentZeroFilledSubarrays = 0

        for num in nums:
            if num != 0:
                cnt = (currentZeroFilledSubarrays*(currentZeroFilledSubarrays+1))//2
                totalZeroFilledSubarrays+=cnt
                currentZeroFilledSubarrays = 0
            else:
                currentZeroFilledSubarrays+=1
        if currentZeroFilledSubarrays != 0:
            cnt = (currentZeroFilledSubarrays*(currentZeroFilledSubarrays+1))//2
            totalZeroFilledSubarrays+=cnt

        return totalZeroFilledSubarrays
        